include/oboe/ResultWithValue.h

```diff
@@ -23,21 +23,61 @@
 
 namespace oboe {
 
+/**
+ * A ResultWithValue can store both the result of an operation (either OK or an error) and a value.
+ *
+ * It has been designed for cases where the caller needs to know whether an operation succeeded and,
+ * if it did, a value which was obtained during the operation.
+ *
+ * For example, when reading from a stream the caller needs to know the result of the read operation
+ * and, if it was successful, how many frames were read. Note that ResultWithValue can be evaluated
+ * as a boolean so it's simple to check whether the result is OK.
+ *
+ * <code>
+ * ResultWithValue<int32_t> resultOfRead = myStream.read(&buffer, numFrames, timeoutNanoseconds);
```

Comment:
Are we recommending:   auto result

Reply:
It's up to the caller to decide whether `auto` is appropriate. In this case I'm explicitly using `ResultWithValue<int32_t>` since the reader may not know that `AudioStream::read` returns that type.